### ISSUE TITLE
Promtail crash: 'panic: assignment to entry in nil map'

### ISSUE THREAD

**Issue by avarsava** [grafana/loki#1659]
**Describe the bug**
Promtail crashed while doing nothing, can't tell why. Restarted and it seems fine

**To Reproduce**
Haven't reproduced

**Expected behavior**
For promtail to run indefinitely

**Environment:**
 - Infrastructure: Ubuntu 18.04
 - Deployment tool: systemd service file

**Screenshots, Promtail config, or terminal output**
```
-- Logs begin at Thu 2020-01-23 11:25:14 UTC, end at Fri 2020-02-07 20:09:00 UTC. --
Feb 06 17:23:01 : systemd[1]: Started Promtail.
Feb 06 17:23:01 : level=info ts=2020-02-06T17:23:01.152858129Z caller=server.go:121 http=[::]:9080 grpc=[::]:39511 msg="server listening on addresses"
Feb 06 17:23:01 : level=info ts=2020-02-06T17:23:01.156664189Z caller=main.go:66 msg="Starting Promtail" version="(version=v1.3.0, branch=HEAD, revision=5e527a03)"
Feb 06 17:23:06 : level=info ts=2020-02-06T17:23:06.152853758Z caller=filetargetmanager.go:257 msg="Adding target" key="{job=\"project\"}"
Feb 06 17:23:06 : ts=2020-02-06T17:23:06.157117649Z caller=log.go:124 component=tailer level=info msg="Seeked /home/ubuntu/logs/project/project.log - &{Offset:0 Whence:0}"
Feb 06 17:23:06 : level=info ts=2020-02-06T17:23:06.160440568Z caller=tailer.go:80 component=tailer msg="start tailing file" path=/home/ubuntu/logs/project/project.log
Feb 06 17:23:16 : panic: assignment to entry in nil map
Feb 06 17:23:16 :         panic: assignment to entry in nil map
Feb 06 17:23:16 : goroutine 185 [running]:
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/positions.(*Positions).PutString(0xc0000a64b0, 0xc000175110, 0x21, 0x25220cc, 0x1)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/positions/positions.go:81 +0xa4
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/positions.(*Positions).Put(0xc0000a64b0, 0xc000175110, 0x21, 0x0)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/positions/positions.go:86 +0x6a
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/targets.(*tailer).markPositionAndSize(0xc0000a9b00, 0x0, 0x0)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/tailer.go:128 +0x176
Feb 06 17:23:16 : github.com/grafana/loki/pkg/helpers.LogError(0x2513fc0, 0x1d, 0xc00006f948)
Feb 06 17:23:16 :         /src/loki/pkg/helpers/logerror.go:10 +0x3c
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/targets.(*FileTarget).run.func1(0xc0005e2080)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/filetarget.go:156 +0x118
Feb 06 17:23:16 : panic(0x2125d00, 0x2a2a910)
Feb 06 17:23:16 :         /usr/local/go/src/runtime/panic.go:679 +0x1b2
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/positions.(*Positions).PutString(0xc0000a64b0, 0xc000175110, 0x21, 0x25220cc, 0x1)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/positions/positions.go:81 +0xa4
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/positions.(*Positions).Put(0xc0000a64b0, 0xc000175110, 0x21, 0x0)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/positions/positions.go:86 +0x6a
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/targets.(*tailer).markPositionAndSize(0xc0000a9b00, 0x0, 0x0)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/tailer.go:128 +0x176
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/targets.(*FileTarget).reportSize(0xc0005e2080, 0xc0000476c0, 0x1, 0x1)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/filetarget.go:330 +0xc3
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/targets.(*FileTarget).sync(0xc0005e2080, 0xc00006fe30, 0x4)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/filetarget.go:216 +0x1d4
Feb 06 17:23:16 : github.com/grafana/loki/pkg/promtail/targets.(*FileTarget).run(0xc0005e2080)
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/filetarget.go:186 +0x602
Feb 06 17:23:16 : created by github.com/grafana/loki/pkg/promtail/targets.NewFileTarget
Feb 06 17:23:16 :         /src/loki/pkg/promtail/targets/filetarget.go:113 +0x24e
Feb 06 17:23:16 systemd[1]: promtail.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Feb 06 17:23:16 systemd[1]: promtail.service: Failed with result 'exit-code'.
```

Positions file:
```
positions:
  /home/ubuntu/logs/project/project.log: "0"
```

**Comment by rfratto** [comment]
I looked into this a little bit and it seems like the only thing that would cause the positions map to be empty is if the positions file had existed at some point but was empty. 

That might not entirely address this, but it should help make it more resilient.

**Comment by owen-d** [comment]
@avarsava @rfratto Can we mark this as closed (at least until it pops back up)?

**Comment by rfratto** [comment]
@owen-d Good idea. Fixed (maybe) by #1660 :man_shrugging:
